Upstream-Status: Pending [Not submitted to upstream yet]
Signed-off-by: Arunachalam Ganapathy <arunachalam.ganapathy@arm.com>

From 3563e39d66404ff2415d29c4018757a85cd0ad97 Mon Sep 17 00:00:00 2001
From: Arunachalam Ganapathy <arunachalam.ganapathy@arm.com>
Date: Wed, 13 Jan 2021 13:49:54 +0000
Subject: [PATCH 1/3] plat-totalcompute: add Theodul DSU support

- Change CPU count from 4 to 8
- Set heap size to 128K
- In debug print, change EMSG to DMSG

Change-Id: Ic426f5e08ac4e7ce1be37333486c75a3228e3415
Signed-off-by: Arunachalam Ganapathy <arunachalam.ganapathy@arm.com>
---
 core/arch/arm/mm/mobj_ffa.c                        |  2 +-
 core/arch/arm/plat-totalcompute/conf.mk            |  4 +++-
 core/arch/arm/plat-totalcompute/optee_manifest.dts |  2 +-
 core/arch/arm/plat-totalcompute/tc0_spmc_pm.c      | 10 +++++++++-
 4 files changed, 14 insertions(+), 4 deletions(-)

diff --git a/core/arch/arm/mm/mobj_ffa.c b/core/arch/arm/mm/mobj_ffa.c
index 8e334f04..500404c3 100644
--- a/core/arch/arm/mm/mobj_ffa.c
+++ b/core/arch/arm/mm/mobj_ffa.c
@@ -381,7 +381,7 @@ struct mobj *mobj_ffa_get_by_cookie(uint64_t cookie,
 		if (mf)
 			DMSG("cookie %#"PRIx64" resurrecting", cookie);
 		if (!mf) {
-			EMSG("Populating mobj from rx buffer\n");
+			DMSG("Populating mobj from rx:  %#"PRIx64"", cookie);
 			mf = thread_spmc_populate_mobj_from_rx(cookie);
 		}
 #endif
diff --git a/core/arch/arm/plat-totalcompute/conf.mk b/core/arch/arm/plat-totalcompute/conf.mk
index d330eb92..c35437aa 100644
--- a/core/arch/arm/plat-totalcompute/conf.mk
+++ b/core/arch/arm/plat-totalcompute/conf.mk
@@ -22,7 +22,9 @@ platform-aflags-debug-info = -gdwarf-2
 endif
 
 ifeq ($(PLATFORM_FLAVOR),tc0)
-CFG_TEE_CORE_NB_CORE = 4
+CFG_TEE_CORE_NB_CORE = 8
+
+CFG_CORE_HEAP_SIZE ?= 131072
 
 ifeq ($(CFG_CORE_SEL1_SPMC),y)
 CFG_TZDRAM_START ?= 0xfd000000
diff --git a/core/arch/arm/plat-totalcompute/optee_manifest.dts b/core/arch/arm/plat-totalcompute/optee_manifest.dts
index 4e60cbe1..60cb5957 100644
--- a/core/arch/arm/plat-totalcompute/optee_manifest.dts
+++ b/core/arch/arm/plat-totalcompute/optee_manifest.dts
@@ -18,7 +18,7 @@
 	ffa-version = <0x00010000>; /* 31:16 - Major, 15:0 - Minor */
 	uuid = <0x486178e0 0xe7f811e3 0xbc5e0002 0xa5d5c51b>;
 	id = <1>;
-	execution-ctx-count = <4>;
+	execution-ctx-count = <8>;
 	exception-level = <2>; /* S-EL1 */
 	execution-state = <0>; /* AARCH64 */
 	load-address = <0xfd280000>;
diff --git a/core/arch/arm/plat-totalcompute/tc0_spmc_pm.c b/core/arch/arm/plat-totalcompute/tc0_spmc_pm.c
index 8e59a8cc..3fb32823 100644
--- a/core/arch/arm/plat-totalcompute/tc0_spmc_pm.c
+++ b/core/arch/arm/plat-totalcompute/tc0_spmc_pm.c
@@ -17,7 +17,11 @@ static const uint64_t core_clus_aff_array[] = {
 	0x0000,		/* Cluster 0 Cpu 0 */
 	0x0001,		/* Cluster 0 Cpu 1 */
 	0x0002,		/* Cluster 0 Cpu 2 */
-	0x0003		/* Cluster 0 Cpu 3 */
+	0x0003,		/* Cluster 0 Cpu 3 */
+	0x0004,		/* Cluster 0 Cpu 4 */
+	0x0005,		/* Cluster 0 Cpu 5 */
+	0x0006,		/* Cluster 0 Cpu 6 */
+	0x0007		/* Cluster 0 Cpu 7 */
 };
 
 void ffa_secondary_cpu_boot_req(vaddr_t secondary_ep, uint64_t cookie)
@@ -38,6 +42,10 @@ void ffa_secondary_cpu_boot_req(vaddr_t secondary_ep, uint64_t cookie)
 		/* Create an mpidr from core_clus_aff_array */
 		a1 |= core_clus_aff_array[cnt] << aff_shift;
 
+		/* Ignore current cpu */
+		if (a1 == mpidr)
+			continue;
+
 		/* Invoke the PSCI_CPU_ON_SMC64 function */
 		ret = thread_smc(PSCI_CPU_ON_SMC64, a1, secondary_ep, cookie);
 
-- 
2.29.2

